[TRTLLM-4517] [feat] Additional model outputs

Funatiq opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Support requesting and returning additional model outputs for both context and generation; exposed on result objects.
  - Advanced quickstart adds --additional_model_outputs to print selected extra outputs.
  - Sampling parameters and LLM now accept additional_model_outputs as a list of strings.

- Improvements
  - More flexible logits handling across models, including encoder-only multimodal scenarios.

- Tests
  - Added unit and integration tests covering additional model outputs.
  - Updated API stability references to include new result fields and parameter types.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
[refactor] Improve logits handling in PyTorchModelEngine

- Added additional checks to ensure proper handling of logits when the model returns a single tensor.
- Added additional checks to ensure proper handling of logits when the model returns a dictionary.
- Adjusted output processing to support both tensor and dictionary formats, ensuring proper gathering of logits based on specified IDs.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16417](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16417/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16671](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16671/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #12513](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12513/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16837](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16837/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #16837](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16837/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #12640](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12640/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16976](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16976/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #16976](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16976/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #12744](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12744/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #17530](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17530/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #17530](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/17530/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13179](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13179/) completed with status: 'FAILURE'

Funatiq committed:
[feat] Add additional model output support to LLM request

- Add additional model output support to LlmRequest initialization.
- Add py_additional_outputs to LlmRequest.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Add additional model output support to PyResult

- Collect additional model outputs in a list in PyResult.
- Concatenate the outputs in the PyResult on access.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Add additional model output support to PyExecutor

- Add HandleAdditionalOutputs class to handle additional model outputs.
- Modify PyExecutor to use HandleAdditionalOutputs.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Enhance quickstart_advanced.py and CompletionOutput for additional model outputs

- Added support for displaying additional context and generation outputs in quickstart_advanced.py.
- Updated CompletionOutput class to include optional fields for additional context and generation outputs.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #18242](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18242/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13672](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13672/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #18622](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18622/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #18622](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18622/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #13980](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/13980/) completed with status: 'FAILURE'

Funatiq committed:
[feat] Introduce DummyModel and tests for additional model outputs

- Added a new model class `DummyModel` with a corresponding configuration class `DummyConfig`.
- Updated `__init__.py` to include `DummyModel` in the module exports.
- Created unit tests for additional model outputs in `SamplingParams`, ensuring proper configuration and integration.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
fixup! [feat] Introduce DummyModel and tests for additional model outputs

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Update DummyModel and integration tests for additional model outputs

- Enhanced the `DummyModel` class to include attention metadata and improved the forward method to return fixed logits, context output, and generation output for testing.
- Modified integration tests to validate the presence and shape of additional model outputs, ensuring compatibility with the new model structure.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[refactor] Move DummyModel to tests

- Deleted the `DummyModel` class and its configuration from modeling.
- Moved the `DummyModel` class to tests.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[ci] Add additional model outputs tests

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
fixup! [feat] Update DummyModel and integration tests for additional model outputs

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A10-PyTorch-1"

Funatiq committed:
[feat] Add additional model output support to quickstart and SamplingParams

- Added command-line arguments for specifying additional model outputs and context gathering in `quickstart_advanced.py`.
- Modified `SamplingParams` to include additional model outputs in the configuration.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #18974](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/18974/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #14224 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/14224/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #20274](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20274/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15287 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15287/) completed with status: 'FAILURE'

Funatiq commented:
/bot run --stage-list "A10-PyTorch-1" --detailed-log

tensorrt-cicd commented:
[PR_Github #20382](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20382/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15377 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15377/) completed with status: 'FAILURE'

Funatiq committed:
test: disable threadleak for additional_model_outputs test

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A10-PyTorch-1" --detailed-log

tensorrt-cicd commented:
[PR_Github #20449](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20449/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20449](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20449/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15416 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15416/) completed with status: 'SUCCESS'

Funatiq committed:
test: use context manager in additional_model_outputs integration test

- Removed the thread leak marker from the integration test.
- Wrapped the output generation in a context manager to ensure proper resource handling.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #20462](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20462/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20522](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20522/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #20522](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20522/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #15480](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/15480/) completed with status: 'FAILURE'

Funatiq committed:
test: improve test for additional outputs

- Add output check for generation output of model output that includes context.
- Make model outputs depend on position ids.
- Verify outputs based on token position.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #20726](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20726/) [ run ] triggered by Bot

Funatiq committed:
refactor: Simplify AdditionalModelOutput

- Remove AdditionalModelOutput class, use str instead (Remove gather_context flag).
- Automatically detect whether to gather the additional outputs for context tokens or not.
- Update tests to reflect the changes.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
refactor: Enhance handling of additional outputs

- Updated the logic to collect additional outputs from context and generation requests.
- Only iterate over the requested additional outputs instead of all tensors.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #20783](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/20783/) [ run ] triggered by Bot

Funatiq commented on tests/unittest/llmapi/test_additional_model_outputs.py:
```diff
@@ -0,0 +1,255 @@
+        AdditionalModelOutput(name="context_output", gather_context=True),
+        AdditionalModelOutput(name="generation_output", gather_context=False),
+    ]
+
```
`gather_context` flag has been removed.

Funatiq committed:
doc: Move additional output parameters to prototype

- Move additional context and generation outputs to prototype in completion output schema.
- Update sampling params schema to set `additional_model_outputs` to prototype.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

DomBrown reviewed (approved):
LGTM regarding the API change.
